proxy.go

```diff
@@ -316,6 +328,13 @@ func (rp *reverseProxy) serveFromCache(s *scope, srw *statResponseWriter, req *h
 		if srw.statusCode != 0 {
 			bufferedRespWriter.WriteHeader(srw.statusCode)
 		}
+
+		// mark transaction as failed
+		// todo: discuss if we should mark it as failed upon timeout. The rational against it would be to hope that
+		// 		 partial results of the query are cached and therefore subsequent execution can succeed
+		if err = userCache.Fail(key); err != nil {
```

Comment:
IMHO you should dissociate timeouted queries from other error reasons. And CHProxy should only deal with timeout erros because let's say there is a transient issue on the cluster (like an AWS network down for 3 sec), all the queries done during this downtime will be unavailable as long as they're in the transaction registry

Reply:
I think it'll be hard to tell the difference between those different types of errors because Clickhouse does not seem to handle it well. I've checked status codes it propagates in case of socket timeout and it's `http service unavailable 503`

https://github.com/ClickHouse/ClickHouse/blob/f0774ca9ee1c6de7049fa10be0c67eeb822797a6/src/Server/HTTPHandler.cpp#L225

Reply:
I'm not sure, I did a test using the http client to get 2 differents behaviour:
- a timeout
- an predicted timeout (so the query is cancel by CH)
I got on both case an HTTP 500 (which is weird because according to the code you showed me, I was exepcted a 503)
but on the first one I got the answer:
Code: 159. DB::Exception: Timeout exceeded: elapsed 1.00059767 seconds, maximum: 1. (TIMEOUT_EXCEEDED) (version 22.2.2.1)
and on the second one I got a 
Code: 160. DB::Exception: Estimated query execution time (220.8800604841333 seconds) is too long. Maximum: 30. Estimated rows to process: 18629040: While executing MergeTreeInOrder. (TOO_SLOW) (version 22.2.2.1)

So maybe we could try to fetch the error code in case of a 500 from CH to check if it's 159 or a 160. WDYT?

Reply:
ok,
since this point is not critical we can leave it the way it is now.